src/main/scala/com/databricks/labs/overwatch/pipeline/GoldTransforms.scala

```diff
@@ -499,122 +501,132 @@ trait GoldTransforms extends SparkSessionWrapper {
+        .withColumn("sum_uptime_in_state_H", sum(col("uptime_in_state_H")).over(aggregationWindow))
+        .withColumn("rn", row_number.over(orderingWindow))
+        .withColumn("rnk", rank.over(orderingWindow))
+        .filter('rnk === 1 && 'rn === 1).drop("rnk", "rn")
+
+      val joinedDF_after_before = joinedDF.filter(state_after_before)
+        .withColumn("sum_total_DBU_cost", sum(col("total_DBU_cost")).over(aggregationWindow))
+        .withColumn("sum_uptime_in_state_H", sum(col("uptime_in_state_H")).over(aggregationWindow))
+        .withColumn("rn", row_number.over(orderingWindow))
+        .withColumn("rnk", rank.over(orderingWindow))
+        .filter('rnk === 1 && 'rn === 1).drop("rnk", "rn")
+
+      val joinedDF_after = joinedDF.filter(state_after)
+        .withColumn("sum_total_DBU_cost", sum(col("total_DBU_cost")).over(aggregationWindow))
+        .withColumn("sum_uptime_in_state_H", sum(col("uptime_in_state_H")).over(aggregationWindow))
+        .withColumn("rn", row_number.over(orderingWindow))
+        .withColumn("rnk", rank.over(orderingWindow))
+        .filter('rnk === 1 && 'rn === 1).drop("rnk", "rn")
+
+      val joinedDF_before_after = joinedDF.filter(state_before_after)
+        .withColumn("sum_total_DBU_cost", sum(col("total_DBU_cost")).over(aggregationWindow))
+        .withColumn("sum_uptime_in_state_H", sum(col("uptime_in_state_H")).over(aggregationWindow))
+        .withColumn("rn", row_number.over(orderingWindow))
+        .withColumn("rnk", rank.over(orderingWindow))
+        .filter('rnk === 1 && 'rn === 1).drop("rnk", "rn")
+
+      val unionDF = joinedDF_before_before.union(joinedDF_after_before).union(joinedDF_after).union(joinedDF_before_after)
+        .withColumn("total_DBU_cost", sum(col("sum_total_DBU_cost")).over(aggregationWindow))
+        .withColumn("uptime_in_state_H", sum(col("sum_uptime_in_state_H")).over(aggregationWindow))
+        .withColumn("rn", row_number.over(orderingWindow))
+        .withColumn("rnk", rank.over(orderingWindow))
+        .drop("sum_total_DBU_cost", "sum_uptime_in_state_H")
+
+
+      val unionDF_final = unionDF.filter('rnk === 1 && 'rn === 1).drop("rnk", "rn")
+      val df = joinedDF.join(unionDF_final, Seq("unixTimeMSStart", "unixTimeMSEnd", "commandId"), "leftAnti")
+      val notebookClsfDF = unionDF_final.union(df.select(unionDF_final.schema.fieldNames.map(col): _*))
+      val notebookCodeAndMetaDF = notebookClsfDF.withColumnRenamed("unixTimeMS_state_start", "unixTimeMS")
+        .toTSDF("unixTimeMS", "organization_id", "date", "notebookId")
+        .lookupWhen(
+          notebookLookupTSDF,
+          maxLookAhead = 100
+        )
+        .df
+        .withColumn("dbu_cost_ps", col("total_DBU_cost") / col("uptime_in_state_H") / lit(3600))
+        .withColumn("estimated_dbu_cost", col("executionTime") * col("dbu_cost_ps"))
+        .drop("cluster_id")
+      notebookCodeAndMetaDF.select(colNames: _*)
+    } else{
```

Comment:
Please move the else case to the beginning and handle the error case before the positive case, otherwise it's a bit hard to get to the else 100+ lines after

Reply:
Done